tools/autograd/gen_python_functions.py

```diff
@@ -483,130 +453,166 @@ def is_noarg_binding(overloads):
-
-    grouped = group_overloads(declarations, is_python_method)
-    is_singleton = len(grouped) == 1
+    traceable = 'true' if all(should_trace(o.function) for o in overloads) else 'false'
 
-    signatures = []
-    dispatch = []
-    for i, dictionary in enumerate(grouped):
-        signature = dictionary['signature']
+    grouped_overloads: Tuple[PythonSignatureGroup, ...] = group_overloads(overloads)
+    is_singleton = len(grouped_overloads) == 1
+    signatures: List[str] = []
+    dispatch: List[str] = []
+    for overload_index, overload in enumerate(grouped_overloads):
+        signature = overload.signature.signature_str()
         signatures.append(f'{cpp_string(str(signature))},')
-        overload_index = i if not is_singleton else None
-        dispatch.append(emit_dispatch_case(overload_index, dictionary, is_python_method))
-
-    if is_singleton:
+        dispatch_body = emit_dispatch_case(overload, namedtuple_typenames)
+        dispatch.append(
+            PY_VARIABLE_CASE.substitute(overload_index=overload_index, body=dispatch_body)
+            if not is_singleton else dispatch_body)
+
+    if noarg:
+        template = PY_VARIABLE_METHOD_NOARGS
+    elif is_singleton:
         template = PY_VARIABLE_METHOD_VARARGS_SINGLETON
     else:
         template = PY_VARIABLE_METHOD_VARARGS
 
-    if module:
-        check_has_torch_function = TORCH_FUNCTION_CHECK.substitute(
-            namespace=NATIVE_NAMESPACE_MAPPING[module],
-            modulename='"' + module + '"',
-            self_="self_" if is_python_method else "nullptr",
-        )
-    else:
-        check_has_torch_function = TORCH_FUNCTION_CHECK.substitute(
-            namespace="THPVariableClass",
-            modulename='"torch.Tensor"',
-            self_="self_" if is_python_method else "nullptr",
-        )
-
-    max_args = max([get_python_argc(decl) for decl in declarations])
-    traceable = 'true' if all(should_trace(d) for d in declarations) else 'false'
-
     return template.substitute(
         name=name,
         pycname=pycname,
         method_header=method_header,
-        max_args=max_args,
+        max_args=max(map(lambda o: o.signature.arguments_count(), overloads)),
         signatures=signatures,
         traceable=traceable,
-        check_has_torch_function=check_has_torch_function,
+        check_has_torch_function=gen_has_torch_function_check(
+            name=name,
+            module=module,
+            noarg=noarg,
+            method=method,
+        ),
         dispatch=dispatch,
         method_footer=method_footer,
-        self_="self_" if is_python_method else "nullptr",
+        self_="self_" if method else "nullptr",
     )
 
+def gen_has_torch_function_check(name: str, module: str, *, noarg: bool, method: bool) -> str:
+    if noarg:
+        return f"""\
+if(check_has_torch_function(self_)) {{
+  return handle_torch_function(self_, "{name}");
+}}
+""" if method else ''
```

Comment:
nit: the `if method` has the same indentation as `if noarg` so I nearly thought it was acutally on the next line haha. Might be easier to read if we just have another nested if block inside

Reply:
done